Improvements in processing redirects with cookie containers.

rassilon opened the pull request:
## Description

Improvements for #2058. This is just a starting point for these improvements, they do as you pointed out need some more tests. I have an outline for Can_Perform_RedirectingPost_With_Receive_Cookies Redirect test, I just need to make it RestSharp only as opposed to calling into the product I'm testing atm.

## Purpose
This pull request is a:

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)


## Checklist

<!-- Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code. -->

- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

alexeyzimarev committed:
WIP

rassilon committed:
Improvements in processing redirects with cookie containers.

rassilon commented:
@dotnet-policy-service agree

rassilon committed:
Added very first of many redirection cookie tests.

rassilon committed:
Added very first of many redirection cookie tests.

rassilon committed:
Update previous redirect test case since, the redirected URL sends cookies as well...

alexeyzimarev commented on src/RestSharp/RestClient.Async.cs:
```diff
@@ -138,23 +153,65 @@ public partial class RestClient {
+                    }
+                }
+
+                // Disallow automatic redirection from secure to non-secure schemes
```
That's one thing I want to be allowed based on redirect options.

rassilon replied:
I'm happy to help do the work if you point me in the general direction of an API surface outline. :)

alexeyzimarev replied:
Check my comment in #2059

rassilon replied:
This specific decision is now driven by the new option object.

rassilon committed:
Additional tests

rassilon committed:
more redirection/cookie related routes

rassilon committed:
Fix build error with respect to System.Web.HttpUtility.ParseQueryString.. wierd one..

rassilon committed:
Add new files..

rassilon committed:
FileParameter: Mark the obsoleted property as NOT CLSCompliant to reduce warnings..
RequestContent: Don't use obsoleted property.
RestClientOptions*: use new RedirectOptions class instead for FollowRedirects.

rassilon commented:
@alexeyzimarev , I'm getting a build error for HttpUtility.ParseQueryString for net472, but heck if I can understand why...

rassilon committed:
Try to pull in the correct assemblies for System.Web.HttpUtility.ParseQueryString

rassilon committed:
Start paying attention to some of the  Options.RedirectionOptions properties.

rassilon committed:
Fix coding style violation..

rassilon commented:
@alexeyzimarev , ok, I've started this in the general direction you mentioned. Still more changes to come. Let me know what you think so far. Should we have a Options.RedirectOptions property that determines the behavior of forcing the method to GET for instance?

rassilon committed:
Improvements (and corrections on some tiny merge errors during) based on rebasing...

rassilon commented:
@alexeyzimarev , I've updated the branch for the interceptor changes. However, should After Request be called once for the last request in the redirect chain or repeatedly for each redirect?

alexeyzimarev commented:
Yeah, I think so. Interceptors don't care about how the request is done. It might change in the future if we decide to add `OnRetry` and `OnRedirect`, but with the current API I think you're right, just call it when it's all done.

rassilon commented:
@alexeyzimarev , one thing we'll need to some of these RedirectOptions is a secure test server... Any thoughts about that fun?

rassilon committed:
Fixes for Interceptor tests.

rassilon committed:
Start the process of testing the new RedirectOptions..

rassilon committed:
Update xunit adapter and visual studio test SDK nuget pkg versions for newer VS versions.

rassilon committed:
ForwardHeaders/ForwardAuthorization/ForwardCookies tests..

rassilon commented:
@alexeyzimarev , I added ForwardHeaders/ForwardAuthorization/ForwardCookies tests... Should ForwardAuthorization support the Authenticator's parameter name instead of just KnownHeaders.Authorization?

rassilon committed:
Add RedirectOptions.ForwardQuery support and tests.

rassilon committed:
Fix Fragment handling with AddQueryString extension, and add ForwardFragment tests.

rassilon committed:
Add RedirectOptions.MaxRedirects support and tests.

rassilon committed:
Add secure test server end point to TestServer.
Add secure test server untrusted certificate.
Out of paranoia, don't let SimpleServer use either the secure or insecure ports from TestServer.
Add tests both positive and negative for RedirectOptions.FollowRedirectsToInsecure

Remember to override options.RemoteCertificateValidationCallback if you get a certificate validation error in new tests.

rassilon committed:
minor tweaks

rassilon commented:
@alexeyzimarev, I've added ForwardQuery, ForwardFragmnet, MaxRedirect, and FollowRedirectsToInsecure related tests.
This leaves ForwardBody (support/test), ForceForwardBody (test), AllowRedirectMethodStatusCodeToAlterVerb (test), and custom redirection status code additions for redirect detection (support and tests).

However, I wonder if the ForceForwardBody support is the best way to achieve that goal... Please take a look and see if you have any feedback.

rassilon commented:
All of the current test explorer tests pass, btw.

rassilon commented:
No problem. I managed to get Visual Studio .Net code coverage analysis working after I was getting nowhere with Coverlet and Fine Code Coverage. I'll be adding the rest of the test cases (indicated by coverage and otherwise), hopefully before the end of the week.

rassilon committed:
RestClient.Async.cs: Convert lots of bools to a [Flag] enum setup.
Add changes to allow the AlterVerb... RedirectOptions to work properly.
RequestContent.cs: Add support for omitting the body (due to the HTTP Verb/Method changing under redirect processing)
TestServer.cs:
* Add certificate password for SSL test server.
* Add HTTP verb changing related routes
* Add dump-request route so that the method and hearders get dumped to response content to help enable authoring RedirectOption tests.
* Minor tweaks to silence nullability warnings.
RedirectTests.cs: Use new (Not)ContainCookieWithNameAndValue extension methods in tests, and use .And to cleanup the repetitive assertion code.
RedirectOptionsTest.cs:
* Add missing StatusCode assertions
* Use new (Not)ContainCookieWithNameAndValue extension methods in the tests and .And. to cleanup the assertions.
* For coverage reasons add initial request cookies to Can_RedirectForwardHeadersFalseWithCookie_DropHeaders.
* New tests:
Options.RedirectOptions.FollowRedirects = false
Options.FollowRedirects = False (back compat)
AllowForcedRedirectVerbChange = false with 302
AllowForcedRedirectVerbChange = false with 303
Change verb with defaults after 302
Change verb with defaults after 303
Don't chanve verb with defaults after 307
Options.CookieContainer contains expected results after a redirection.
Change verb with defaults after 303, but with ForceForwardBody so that the request body is forwarded on the new verb.

Additionally, due to having VS 17.9.3 updated xunit/ms test sdk nuget pkgs that makes some of the dependabot PRs obsolete.

rassilon commented:
@alexeyzimarev, that's the last set of changes I will make before we discuss ForwardBody vs ForceForwardBody. If we should only have, say, ForceForwardBody since RedirectMethod status code (303) returns true from the helper method to determine if we should force GET instead of the current method.

I don't anticipate making any other changes before you get a chance to give me some feedback.

The code coverage looks excellent. I noticed that sometimes some of the non-"TestServer.cs" related tests fail due to trying to dispose of an already disposed server. I don't know if this is related to the xunit nuget pkg update. I also don't know if moving to IAsyncLifetime like TestServer.cs would fix them.

rassilon commented:
Bah, I just had this thought:
Should we ALWAYS regenerate the Cookie header after a redirect for security reasons?
And if so, should we even have ForceCookie? I'm leaning towards no.

For non-RestSharp users, if a supplied HttpClientHandler sets a CookieContainer, the HttpClient code does this as near as I can unravel from the source code. (That code makes my head hurt....) The basic idea is that the System.Net.Http inner handler of the RedirectHandler passes doAuth = false to SendAsync overloads of various sorts.

However, RestSharp doesn't supply the CookieContainer to a HttpClient HttpClientHandler, so we have to do it ourselves.

alexeyzimarev commented on src/RestSharp/Options/RestClientOptions.cs:
```diff
@@ -60,7 +66,7 @@ public class RestClientOptions {
             : ResponseStatus.Error;
 
-    /// <summary>
+    /// <summary>s
```
I guess this `s` needs to go?

rassilon replied:
Yep. Fixed locally.

alexeyzimarev commented:
I am not sure. I thought we should collect the response headers and pass them over to the next call. It might, indeed, create security concerns if the response contains things like authentication cookie.
I didn't look at HttpClient code to see how they handle redirects. Instead, I checked Refit and I thought we could/should do the same thing, or similar. I am wondering if you are making it too good? :)

rassilon commented:
Hah. I think removing ForceCookie, regenerating the cookies for every redirected request (for cross-domain security reasons), and preventing the verb change without supporting ForwardBody or ForceForwardBody nonsense would be good enough. There's a reason the CookieContainer has you supply the requesting Uri for the GetCookie method after all. Hrm.. I looked at Refit briefly (https://github.com/reactiveui/refit/) but didn't see the bits that did header propagation. Did I look in the wrong place?

rassilon commented:
We could spin up a slack restsharp instance and have a conference call to discuss in a little more detail if you want or something similar...

alexeyzimarev commented:
Maybe create a Discord server?

rassilon commented:
> Maybe create a Discord server?

@alexeyzimarev , https://discord.com/invite/MrGRGUug (expires in 7 days)
